snap-update-tools/check_snap_availability.py

```diff
@@ -0,0 +1,218 @@
+    required-snaps:
+        - name: checkbox22
+            channels:
+                - edge
+            architectures:
+                - amd64
+                - arm64
+
+"""
+
+import argparse
+import requests
+import time
+import yaml
+from dataclasses import dataclass
+
+
+# the dataclass for the concrete snap specification,
+# instance of this class represents one, concrete snap
+@dataclass
+class SnapSpec:
+    name: str
+    version: str
+    channel: str
+    arch: str
+
+    def __hash__(self) -> int:
+        # needed so we can this dataclass instances as keys
+        return hash((self.name, self.version, self.channel, self.arch))
+
+
+def query_store(snap_spec: SnapSpec) -> dict:
+    """
+    Pull the information about the snap from the snap store.
+    :param snap_spec: the snap specification
+    :return: deserialised json with the response from the snap store
+    """
+    # the documentation for this API is at https://api.snapcraft.io/docs/search.html
+    url = "https://api.snapcraft.io/v2/snaps/find"
+    headers = {"Snap-Device-Series": "16", "Snap-Device-Store": "ubuntu"}
+    params = {
+        "q": snap_spec.name,
+        "channel": snap_spec.channel,
+        "fields": "revision,version",
+        "architecture": snap_spec.arch,
+    }
+
+    response = requests.get(url, headers=headers, params=params)
+    return response.json()
+
+
+def get_snap_info_from_store(snap_spec: SnapSpec) -> dict:
+    """
+    Get detailed information about a snap using the info endpoint.
+
+    :param snap_spec: the snap specification
+    :return: deserialised json with the response from the snap store
+    """
+    url = f"https://api.snapcraft.io/v2/snaps/info/{snap_spec.name}"
+    headers = {"Snap-Device-Series": "16", "Snap-Device-Store": "ubuntu"}
+    response = requests.get(url, headers=headers)
+    if response.status_code != 200:
+        raise SystemExit(
+            f"Failed to get info about {snap_spec.name} from the snap store."
+        )
+
+    return response.json()
+
+
+def is_snap_available(snap_spec: SnapSpec, store_response: dict) -> bool:
+    """
+    Process the response from the snap store and check whether the specified
+    snap is available.
+    :param snap_spec: the snap specification
+    :param store_response: the response from the snap store
+    :return: True if the snap is available, False otherwise
+    """
+    # let's split the "channel" provided by the user
+    # into the track and the risk
+
+    def matches_spec(channel_info: dict) -> bool:
+        # some of the channel names don't follow the "track/risk" format
+        # so we need to handle them separately
+        if "/" not in snap_spec.channel:
+            return (
+                channel_info["channel"]["name"] == snap_spec.channel
+                and channel_info["version"] == snap_spec.version
+                and channel_info["channel"]["architecture"] in snap_spec.arch
+            )
+        else:
```

Comment:
can be removed so that we get
```python
if exceptional_case:
  return exceptional_result
return normal_result
```

Reply:
this branching exhausts the list of possibilities, and neither of the branches is more exceptional than others. It's not like the first return is an "early return". And in cases like this I put emphasis on that with contrasting if/else. Sort of like in a stronger and more formal languages.